provider/domeneshop/domeneshop.go

```diff
@@ -0,0 +1,311 @@
+}
+
+// NewDomeneshopProvider initializes a new Domeneshop DNS based Provider.
+func NewDomeneshopProvider(ctx context.Context, domainFilter endpoint.DomainFilter, dryRun bool, appVersion string) (*DomeneshopProvider, error) {
+	apiToken, ok := os.LookupEnv("DOMENESHOP_API_TOKEN")
+	if !ok {
+		return nil, fmt.Errorf("no api token found")
+	}
+
+	apiSecret, ok := os.LookupEnv("DOMENESHOP_API_SECRET")
+	if !ok {
+		return nil, fmt.Errorf("no api secret found")
+	}
+
+	client := NewClient(apiToken, apiSecret, appVersion)
+
+	provider := &DomeneshopProvider{
+		Client:       client,
+		domainFilter: domainFilter,
+		DryRun:       dryRun,
+	}
+	return provider, nil
+}
+
+// Records returns the list of records in a given zone.
+func (p *DomeneshopProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	var endpoints []*endpoint.Endpoint
+	domains, err := p.Client.ListDomains(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, domain := range domains {
+		records, err := p.Client.ListDNSRecords(ctx, domain, "", "")
+		if err != nil {
+			return nil, err
+		}
+
+		for _, record := range records {
+			dnsName := record.Host + "." + domain.Name
+
+			// root name is identified by @ and should be
+			// translated to zone name for the endpoint entry.
+			if record.Host == "@" {
+				dnsName = domain.Name
+			}
+
+			e := endpoint.NewEndpointWithTTL(dnsName, record.Type, endpoint.TTL(record.TTL), record.Data)
+
+			endpoints = append(endpoints, e)
+		}
+	}
+
+	log.WithFields(log.Fields{
+		"endpoint_count": len(endpoints),
+	}).Info("Found endpoints.")
+
+	return mergeEndpointsByNameType(endpoints), nil
+}
+
+func recordFromEndpointTarget(domain *Domain, e *endpoint.Endpoint, target string) *DNSRecord {
+	record := &DNSRecord{
+		Type: e.RecordType,
+		Host: strings.TrimSuffix(e.DNSName, "."+domain.Name),
+		Data: target,
+		TTL:  getTTLFromEndpoint(e),
+	}
+
+	if record.Host == domain.Name {
+		record.Host = "@"
+	}
+
+	return record
+}
+
+// ApplyChanges applies a given set of changes in a given zone.
+func (p *DomeneshopProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	domains, err := p.Client.ListDomains(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, domain := range domains {
+		for _, change := range changes.Create {
+			if strings.HasSuffix(change.DNSName, domain.Name) {
+				log.WithFields(log.Fields{
+					"endpoint": change,
+				}).Info("Creating endpoint")
+
+				for _, target := range change.Targets {
+					record := recordFromEndpointTarget(domain, change, target)
+
+					log.WithFields(log.Fields{
+						"record": record,
+					}).Info("Creating record")
+
+					err := p.Client.AddDNSRecord(ctx, domain, record)
+					if err != nil {
+						var httpErr *HttpError
+						if errors.As(err, &httpErr) {
+							if httpErr.Response.StatusCode == 409 && httpErr.ErrorBody.Code == "record:collision" {
+								// collision is ok, usually because a record exists without a corresponding registry entry
```

Comment:
This happened a few times in my testing:

- Service starts
- Service wants to create CNAME, TXT, TXT records.
- CNAME record is created before the TXT records
- Service crashes
- Service restarts
- Service wants to create CNAME, TXT, TXT records. (since no TXT records exist)
- Collision on the CNAME record.

If there is a better canonical way of handling this let me know :)